### PR TITLE
feat: optimize truncate(col) == value to startsWith predicate

### PR DESCRIPTION
**Pull request by shangxinli**
Rewrite truncate(col) == "value" predicates to col STARTS_WITH "value" for string columns. This enables better predicate pushdown to storage formats and efficient use of prefix indexes.

The optimization only applies when:
- Operation is equality
- Term is a truncate transform
- Literal is a string type

Added tests to verify correct application and edge cases.

### CODE REVIEW

**Comment by wgtmac** [comment]
There are some merge conflicts...

**Comment by shangxinli** [comment]
Fixed

File: src/iceberg/expression/predicate.cc [hunk 1]
```diff
@@ -143,6 +145,27 @@ bool IsFloatingType(TypeId type) {
   return type == TypeId::kFloat || type == TypeId::kDouble;
 }
 
+/// \brief Count the number of UTF-8 code points in a string.
+/// This matches the behavior of TruncateUtils::TruncateUTF8.
+///
+/// NOTE: This counts code points, not grapheme clusters (user-perceived characters).
+/// Per the Iceberg spec, combining marks count as separate code points.
```
Comment (wgtmac) [inline src/iceberg/expression/predicate.cc]: Iceberg spec does not have this. We should not mislead readers.

File: src/iceberg/expression/predicate.cc [hunk 2]
```diff
@@ -143,6 +145,27 @@ bool IsFloatingType(TypeId type) {
...
+/// Example: "é" as e + combining-acute (U+0065 U+0301) = 2 code points,
+///          but "é" as single precomposed character (U+00E9) = 1 code point.
+///
+/// \param str The UTF-8 encoded string
+/// \return The number of code points (not bytes, not graphemes)
+inline int32_t CountUTF8CodePoints(std::string_view str) {
```
Comment (wgtmac) [inline src/iceberg/expression/predicate.cc]: Perhaps this should be moved to `string_util.h` or `truncate_util.h`

File: src/iceberg/expression/predicate.cc [hunk 3]
```diff
@@ -246,7 +269,69 @@ Result<std::shared_ptr<Expression>> UnboundPredicate<B>::BindLiteralOperation(
...
+  //   implies exact match, so STARTS_WITH remains valid (short-string invariance)
+  if (BASE::op() == Expression::Operation::kEq &&
+      bound_term->kind() == Term::Kind::kTransform) {
+    // Safe to cast after kind check confirms it's a transform
+    auto* transform_term = dynamic_cast<BoundTransform*>(bound_term.get());
+    if (!transform_term) {
+      // Should never happen after kind check, but be defensive
+      return std::make_shared<BoundLiteralPredicate>(BASE::op(), std::move(bound_term),
+                                                     std::move(literal));
+    }
```
Comment (wgtmac) [inline src/iceberg/expression/predicate.cc]:
```suggestion
    auto* transform_term = internal::checked_cast<BoundTransform*>(bound_term.get());
```

Let's reuse the existing pattern for cast.

File: src/iceberg/expression/predicate.cc [hunk 4]
```diff
@@ -246,7 +269,69 @@ Result<std::shared_ptr<Expression>> UnboundPredicate<B>::BindLiteralOperation(
...
+
+    if (transform_term->transform()->transform_type() == TransformType::kTruncate &&
+        literal.type()->type_id() == TypeId::kString &&
+        !literal.IsNull()) {  // Null safety: skip null literals
+
+      // Extract width parameter using type-safe API
+      auto width_opt = transform_term->transform()->param();
```
Comment (wgtmac) [inline src/iceberg/expression/predicate.cc]: I think we can greatly simplify this line and below.

We can add `transform_func` like below.

```cpp
class ICEBERG_EXPORT BoundTransform : public BoundTerm {
 public:
  const std::shared_ptr<TransformFunction>& transform_func() const { return transform_func_; }
};
```

Then we just need to call `transform_term->transform_func()->Transform(literal)` and check if its return value is same as `literal`. We don't even need `CountUTF8CodePoints` above.

File: src/iceberg/transform.h [hunk 5]
```diff
@@ -141,6 +141,19 @@ class ICEBERG_EXPORT Transform : public util::Formattable {
   /// \brief Returns the transform type.
   TransformType transform_type() const;
 
+  /// \brief Returns the optional parameter for parameterized transforms.
+  ///
+  /// For transforms like bucket(N) or truncate(W), returns the parameter value.
+  /// For non-parameterized transforms (identity, year, etc.), returns std::nullopt.
+  ///
+  /// \return The parameter if present, otherwise std::nullopt
+  std::optional<int32_t> param() const {
```
Comment (wgtmac) [inline src/iceberg/transform.h]: Can we remove this function? This makes it harder to evolve the variant in the future.

File: src/iceberg/expression/predicate.cc (hunk 1 above)
Comment (shangxinli) [reply src/iceberg/expression/predicate.cc]: Removed

File: src/iceberg/expression/predicate.cc (hunk 2 above)
Comment (shangxinli) [reply src/iceberg/expression/predicate.cc]: moved

File: src/iceberg/expression/predicate.cc (hunk 3 above)
Comment (shangxinli) [reply src/iceberg/expression/predicate.cc]: sure

File: src/iceberg/expression/predicate.cc (hunk 4 above)
Comment (shangxinli) [reply src/iceberg/expression/predicate.cc]: sure

File: src/iceberg/transform.h (hunk 5 above)
Comment (shangxinli) [reply src/iceberg/transform.h]: sure

File: src/iceberg/expression/predicate.cc [hunk 6]
```diff
@@ -286,7 +289,52 @@ Result<std::shared_ptr<Expression>> UnboundPredicate<B>::BindLiteralOperation(
...
+                                         std::move(literal));
+    }
+
+    if (transform_term->transform()->transform_type() == TransformType::kTruncate &&
+        literal.type()->type_id() == TypeId::kString &&
+        !literal.IsNull()) {  // Null safety: skip null literals
+
+      // Apply truncate transform to the literal and check if result matches
+      // This verifies the literal is compatible with the truncate operation
+      auto transformed_result = transform_term->transform_func()->Transform(literal);
```
Comment (zhjwpku) [inline src/iceberg/expression/predicate.cc:307]: I'm wondering whether we can simply count the code points in the literal and compare that with `TruncateTransform.width()`. If they match, we can rewrite it as a `StartsWith`. We don't need to transform the extended_literal then.

Comment (zhjwpku) [reply src/iceberg/expression/predicate.cc:307]: Also, if the number of code points in the literal is greater than `TruncateTransform.width()`, we can rewrite it as `AlwaysFalse`. I created PR [1] for the Java implementation to do this.

[1] https://github.com/apache/iceberg/pull/16383
